d1985r0.md

```diff
@@ -414,6 +423,163 @@ reflection.
+using x = puts_one<takes_var>; // Error! ... behaves covariantly!
+using w = puts_var<takes_one>; // OK, compiles
+```
+
+Turns out parameter packs behave *covariantly* - exactly the opposite of what
+one would expect given the concepts example above.
+
+We have inconsistent behavior across the language. Let's say we did the right
+thing and made `template auto` behave contravariantly, like concepts.
+
+```cpp
+template <template <auto> typename f> using puts_value = void;
+template <template <template auto> typename f> using puts_any = void;
+template <auto> using takes_value = void;
+template <template auto> using takes_any = void;
+using x = puts_value<takes_any>; // OK
+using w = puts_any<takes_value>; // Error because contravariant.
+```
+
+But then, how do we write `apply`? Let's do it for a single argument to avoid
+complications with a covariant `...`:
+
+```cpp
+template <template <template auto T> typename f, template auto arg>
+using apply1 = f<arg>;
+```
+
+The above is correct, but _useless_ - it requires `f`s signature to be
+`template <template auto T>`. What we want to express is that `f` is
+any kind of unary template template, so we can pass in something like
+`template <int x> using int_constant = std::integral_constant<int, x>;`.
+
+Let's call the covariant version of `template auto` (with the meaning "deduce
+this from the argument") `auto template`:
+
+```cpp
+template <template <auto template T> typename f, template auto arg>
+using apply1 = f<arg>;
+```
+
+This is what we want to express (and check at instantiation time), but now
+the `arg`s constraint is spelled differently from `f`'s constraint, and that
+might be very, very difficult to teach.
+
+It also requires us to reserve an additional combination of tokens. Contrast
+what happens if we just made `template auto` behave covariantly (the way
+`auto template` behaves above) if used that place. What do we lose?
+
+At first glance, we lose the ability to define the contravariant meaning of
+`template auto` - a template parameter that *can* bind to anything. But can
+we get that back?
+
+Recall that concepts behave contravariantly. Consider this one:
+
+```cpp
+namespace std {
+  template <template auto arg>
+  concept anything = true;
+};
+```
+
+If we relaxed the rule that only type concepts could appear in the template
+parameter list, we could say this:
+
+```cpp
+template <template <std::anything ARG> typename takes_anything, template auto arg>
+using apply1 = takes_anything<arg>;
+```
+
+This would work *contravariantly*! While `template auto ARG` means "deduce",
+a function taking a concept (which behaves contravariantly!) that *anything*
+satisfies *has* to be a function taking `template auto` (or `std::anything`).
+
+In light of this, the paper authors have come to the conclusion that trying
+to make `template auto` behave contravariantly is all downside and little
+upside. The example to follow with `template auto` should be the behavior of
```

Comment:
You should mention that as of C++17 this is also supported by the fact that it is allowed to use non-variadic actuals where variadic formals were specified. Not allowing this here would just be non-orthogonal (in addition to useless).

Reply:
Agreed. Expanded wording.